engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala

```diff
@@ -271,48 +268,49 @@ trait ProgramExecutionSupport {
     * @param ctx the runtime context
     * @return the API message describing the error
     */
-  private def getExecutionOutcome(
+  def getExecutionOutcome(
     t: Throwable
-  )(implicit ctx: RuntimeContext): Option[Api.ExecutionResult] = {
-    t match {
-      case ex: AbstractTruffleException
-          if Option(ctx.executionService.getLanguage(ex))
-            .forall(_ == LanguageInfo.ID) =>
-        val section = Option(ctx.executionService.getSourceLocation(ex))
-        Some(
-          Api.ExecutionResult.Diagnostic.error(
-            ex.getMessage,
-            section.flatMap(sec => findFileByModuleName(sec.getSource.getName)),
-            section.map(LocationResolver.sectionToRange),
-            section
-              .flatMap(LocationResolver.getExpressionId(_))
-              .map(_.externalId),
-            ErrorResolver.getStackTrace(ex)
-          )
-        )
+  )(implicit ctx: RuntimeContext): Option[Api.ExecutionResult] =
+    getDiagnosticOutcome.orElse(getFailureOutcome).lift(t)
 
-      case ex: ConstructorNotFoundException =>
-        Some(
-          Api.ExecutionResult.Failure(
-            ex.getMessage,
-            findFileByModuleName(ex.getModule)
-          )
-        )
+  /** Extract diagnostic information from the provided exception. */
+  def getDiagnosticOutcome(implicit
+    ctx: RuntimeContext
+  ): PartialFunction[Throwable, Api.ExecutionResult.Diagnostic] = {
+    case ex: AbstractTruffleException
+        if Option(ctx.executionService.getLanguage(ex))
```

Comment:
It's not directly relevant to this PR (as this was here before), but I was wondering (and it doesn't seem to be documented) - what is the purpose of this guard? It lets through `AbstractTruffleException`s if the language is null or is Enso - so it specifically ignores errors from different languages, unless language is not set at all.
It seems a bit weird to me, because later we look for the stacktrace so it seems maybe we'd want to have *only Enso* here? So I'm wondering if it wasn't supposed to be `exists` instead of `forall`, but I cannot judge that without documentation that would clarify the intent behind this.

Reply:
The guard filters only the Enso errors. But currently, `getLanguage` can only tell the language by the source file, and return `null` when the error originates in built-in nodes. It means that we need `forall` to include builtins as well. Builtins cover a significant part of the language, and most of the calls end up there, so we definitely want to support them.

The guard is definitely not obvious, so I'll add a comment there :+1:

Reply:
> The guard filters only the Enso errors. But currently, `getLanguage` can only tell the language by the source file, and return `null` when the error originates in built-in nodes.

So I understand that technically we can get a node with `null` language that is not from Enso, right? But we have no good way to check that? If yes, if we get such input, do we expect that it could crash or is it safe to execute the rest of the code on it? (And if it is safe, why do we ignore other languages then?)

Reply:
It should be safe to execute the code on any exception. When the code was written, we only had an Enso and a host (Java) languages, and the idea was to provide the right context and show only stuff related to Enso